### ISSUE TITLE
ApplicationCable::Connection has been removed from the module tree but is still active

### ISSUE THREAD

**Issue by Sen** [anycable/anycable-rails#82]
**Ruby version: 2.5.1**

**Rails version: 5.2.2**

**`anycable` gem version: 0.6.1**

**`anycable-rails` gem version: 0.6.2**

**`grpc` gem version: grpc-1.18.0-universal-darwin**

in Dev mode, when i leave it for a few minutes, these error coming up:

``` log
13:27:42 cable.1              | [AnyCable sid=WlNSBtRTBwb90dwd_GxnoG] A copy of ApplicationCable::Connection has been removed from the module tree but is still active!:
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/dependencies.rb:495:in `load_missing_constant'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/core_ext/active_support.rb:58:in `block in load_missing_constant'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/core_ext/active_support.rb:16:in `allow_bootsnap_retry'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/bootsnap-1.3.2/lib/bootsnap/load_path_cache/core_ext/active_support.rb:57:in `load_missing_constant'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/dependencies.rb:195:in `const_missing'
13:27:42 cable.1              | /Users/sen/code/apps/beachpricex/app/channels/application_cable/connection.rb:47:in `decoded_auth_token'
13:27:42 cable.1              | /Users/sen/code/apps/beachpricex/app/channels/application_cable/connection.rb:23:in `authorize_cable_request'
13:27:42 cable.1              | /Users/sen/code/apps/beachpricex/app/channels/application_cable/connection.rb:9:in `connect'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/anycable-rails-0.6.2/lib/anycable/rails/actioncable/connection.rb:54:in `handle_open'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/anycable-0.6.1/lib/anycable/rpc_handler.rb:24:in `connect'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/anycable-0.6.1/lib/anycable/handler/capture_exceptions.rb:22:in `block in connect'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/anycable-0.6.1/lib/anycable/handler/capture_exceptions.rb:28:in `capture_exceptions'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/anycable-0.6.1/lib/anycable/handler/capture_exceptions.rb:22:in `connect'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/grpc-1.18.0-universal-darwin/src/ruby/lib/grpc/generic/rpc_desc.rb:60:in `call'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/grpc-1.18.0-universal-darwin/src/ruby/lib/grpc/generic/rpc_desc.rb:60:in `block in handle_request_response'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/grpc-1.18.0-universal-darwin/src/ruby/lib/grpc/generic/interceptors.rb:178:in `block (2 levels) in intercept!'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/grpc-1.18.0-universal-darwin/src/ruby/lib/grpc/generic/interceptors.rb:181:in `block in intercept!'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/anycable-0.6.1/lib/anycable/middleware.rb:11:in `block in request_response'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/anycable-rails-0.6.2/lib/anycable/rails/middlewares/log_tagging.rb:16:in `block in call'
13:27:42 cable.1              | /Users/sen/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-5.2.2/lib/active_support/tagged_logging.rb:71:in `block in tagged'
```

this is strange.

**Comment by Envek** [comment]
I see it too sometimes. It is most probably due to Rails code reloading in dev mode. Haven't investigated yet, but as a workaround you can alway restart RPC process in your process manager (would it be [Overmind](https://github.com/DarthSim/overmind), Foreman, docker-compose or whatever):
```sh
overmind restart cable
```
```sh
docker-compose restart cable
```

**Comment by nixme** [comment]
Could we keep the issue open until we understand what's wrong with the Executor middleware https://github.com/anycable/anycable-rails/blob/master/lib/anycable/rails/middlewares/executor.rb )?

Or I can create a new issue if the moderators prefer.

**Comment by palkan** [comment]
> Could we keep the issue open until we understand what's wrong with the Executor middleware

Agree.

**Comment by bingjyang** [comment]
Janky, but worked around this for now by putting `ApplicationCable::Connection` in an non-autoloaded path.

**Comment by sponomarev** [comment]
@bingjyang Can you show the way how you do it? Did you add `ApplicationCable::Connection` to autoloaded path explicitly before?

**Comment by sponomarev** [comment]
I'm concened because according to the default paths,  `ApplicationCable::Connection` is outside of autoload lookup.

https://github.com/rails/rails/blob/4aa6c328847bd42776f75befefbcc61cd1ec9f20/railties/lib/rails/engine/configuration.rb#L41-L65

**Comment by bingjyang** [comment]
We happened to have it in an auto loaded path, which it probably shouldn’t
have been. So maybe that is the root of the problem.

**Comment by bingjyang** [comment]
Getting back to this... I suppose we followed the recommendation [here](https://guides.rubyonrails.org/action_cable_overview.html#connection-setup) and had it in `app/channels/` (or maybe `rails new` auto-created it). Hence, it was being auto-loaded.

**Comment by sponomarev** [comment]
For all interested, should be fixed in #85

**Comment by bingjyang** [comment]
Thank you!

**Comment by palkan** [comment]
Closing since #85 has been released in 0.6.3
